File: src/websocket.zig
//! WebSocket client implementation for jok framework.
//!
//! This module provides WebSocket support for WebAssembly, it uses the browser's
//! native WebSocket API via JavaScript interop.
//!
//! The WebSocket operates in binary mode (binaryType = 'arraybuffer'), which is the
//! standard for game development as it provides better performance and efficiency for
//! binary data transmission.
//!
//! ## Usage
//!
//! ```zig
//! const jok = @import("jok");
//!
//! var ws: *jok.WebSocket = undefined;
//!
//! pub fn init(ctx: jok.Context) !void {
//!     ws = try jok.WebSocket.create(ctx.allocator(), "ws://localhost:8080");
//!     _ = try ws.on_message.connect(onWebSocketMessage, .{});
//!     _ = try ws.on_close.connect(onWebSocketClose, .{});
//! }
//!
//! fn onWebSocketMessage(websocket: *jok.WebSocket, data: []const u8) void {
//!     // Handle received binary message
//! }
//!
//! fn onWebSocketClose(websocket: *jok.WebSocket) void {
//!     // Handle connection close
//! }
//! ```
//!
//! Written by Claude Sonnet 4.5, reviewed by Jack-Ji

const std = @import("std");
const builtin = @import("builtin");
const signal = @import("utils/signal.zig");

/// Signal types for WebSocket events
pub const OpenSignal = signal.Signal(&.{*WebSocket});
pub const MessageSignal = signal.Signal(&.{ *WebSocket, []const u8 });
pub const ErrorSignal = signal.Signal(&.{*WebSocket});
pub const CloseSignal = signal.Signal(&.{*WebSocket});
var sockets: std.AutoHashMapUnmanaged(usize, *WebSocket) = .{};

/// WebSocket client handle
pub const WebSocket = struct {
    handle: usize,
    allocator: std.mem.Allocator,
    on_open: *OpenSignal,
    on_message: *MessageSignal,
    on_error: *ErrorSignal,
    on_close: *CloseSignal,

    /// Create a new WebSocket connection
    pub fn create(allocator: std.mem.Allocator, url: []const u8) !*WebSocket {
        if (!builtin.cpu.arch.isWasm()) {
            return error.WebSocketNotSupported;
        }

        const ws = try allocator.create(WebSocket);
        errdefer allocator.destroy(ws);
        ws.* = .{
            .handle = 0,
            .allocator = allocator,
            .on_open = OpenSignal.create(allocator) catch unreachable,
            .on_message = MessageSignal.create(allocator) catch unreachable,
            .on_error = ErrorSignal.create(allocator) catch unreachable,
            .on_close = CloseSignal.create(allocator) catch unreachable,
        };
        errdefer {
            ws.on_open.destroy();
            ws.on_message.destroy();
            ws.on_error.destroy();
            ws.on_close.destroy();
        }

        const ws_ptr = @intFromPtr(ws);
        ws.handle = wasm_websocket_create(url.ptr, url.len, ws_ptr);
        if (ws.handle == 0) {
            return error.WebSocketCreateFailed;
        }
        errdefer wasm_websocket_destroy(ws.handle);

        try sockets.put(allocator, ws.handle, ws);
        return ws;
    }

    /// Send a binary message through the WebSocket
    pub fn sendMessage(self: *WebSocket, data: []const u8) void {
        if (builtin.cpu.arch.isWasm()) {
            wasm_websocket_send(self.handle, data.ptr, data.len);
        } else {
            @panic("WebSocket is only supported on WebAssembly platform");
        }
    }

    /// Destroy the WebSocket connection
    pub fn destroy(self: *WebSocket) void {
        if (builtin.cpu.arch.isWasm()) {
            _ = sockets.remove(self.handle);
            wasm_websocket_destroy(self.handle);
            self.on_open.destroy();
            self.on_message.destroy();
            self.on_error.destroy();
            self.on_close.destroy();
            self.allocator.destroy(self);
        } else {
            @panic("WebSocket is only supported on WebAssembly platform");
        }
    }
};

// WebAssembly JavaScript interop functions
extern fn wasm_websocket_create(url: [*]const u8, url_len: usize, ws_ptr: usize) usize;
extern fn wasm_websocket_send(handle: usize, data: [*]const u8, len: usize) void;
extern fn wasm_websocket_destroy(handle: usize) void;

/// Called by JavaScript when the connection is opened
export fn jok_websocket_on_open(handle: usize) void {
    const ws = sockets.get(handle) orelse return;
    ws.on_open.emit(.{ws});
}

/// Called by JavaScript when a message is received
export fn jok_websocket_on_message(handle: usize, data_ptr: [*]const u8, data_len: usize) void {
    const ws = sockets.get(handle) orelse return;
    const data = data_ptr[0..data_len];
    ws.on_message.emit(.{ ws, data });
}

/// Called by JavaScript when an error occurs
export fn jok_websocket_on_error(handle: usize) void {
    const ws = sockets.get(handle) orelse return;
    ws.on_error.emit(.{ws});
}

/// Called by JavaScript when the connection is closed
export fn jok_websocket_on_close(handle: usize) void {
    const ws = sockets.get(handle) orelse return;
    ws.on_close.emit(.{ws});
}

File: src/entrypoints/websocket.js
// WebSocket JavaScript library for jok framework
// This file provides the JavaScript implementation of WebSocket functionality
// that is called from Zig code via extern functions.
//
// Written by Claude Sonnet 4.5, reviewed by Jack-Ji

mergeInto(LibraryManager.library, {
    wasm_websocket_create: function(url_ptr, url_len, ws_ptr) {
        try {
            const url = UTF8ToString(url_ptr, url_len);
            const ws = new WebSocket(url);
            ws.binaryType = 'arraybuffer';

            if (!Module._websockets) {
                Module._websockets = {};
                Module._websocket_next_id = 1;
            }

            const id = Module._websocket_next_id++;
            Module._websockets[id] = ws;

            ws.onopen = function() {
                try {
                    Module.ccall('jok_websocket_on_open', null, ['number'], [id]);
                } catch (e) {
                    console.error('Error calling jok_websocket_on_open:', e);
                }
            };

            ws.onmessage = function(event) {
                const data = new Uint8Array(event.data);
                const data_ptr = _malloc(data.length);
                try {
                    HEAPU8.set(data, data_ptr);
                    Module.ccall('jok_websocket_on_message', null, ['number', 'number', 'number'], [id, data_ptr, data.length]);
                } finally {
                    _free(data_ptr);
                }
            };

            ws.onerror = function(error) {
                console.error('WebSocket error:', error);
                try {
                    Module.ccall('jok_websocket_on_error', null, ['number'], [id]);
                } catch (e) {
                    console.error('Error calling jok_websocket_on_error:', e);
                }
            };

            ws.onclose = function() {
                try {
                    Module.ccall('jok_websocket_on_close', null, ['number'], [id]);
                } catch (e) {
                    console.error('Error calling jok_websocket_on_close:', e);
                }
                delete Module._websockets[id];
            };

            return id;
        } catch (e) {
            console.error('Failed to create WebSocket:', e);
            return 0;
        }
    },

    wasm_websocket_send: function(handle, data_ptr, len) {
        if (!Module._websockets) return;

        const ws = Module._websockets[handle];
        if (ws && ws.readyState === WebSocket.OPEN) {
            const data = HEAPU8.slice(data_ptr, data_ptr + len);
            ws.send(data.buffer);
        }
    },

    wasm_websocket_destroy: function(handle) {
        if (!Module._websockets) return;

        const ws = Module._websockets[handle];
        if (ws) {
            ws.onopen = null;
            ws.onmessage = null;
            ws.onerror = null;
            ws.onclose = null;
            delete Module._websockets[handle];
            ws.close();
        }
    }
});
